cdap-api/src/main/java/co/cask/cdap/api/DatasetConfigurer.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.data.stream.Stream;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.module.DatasetModule;
+
+/**
+ * Provides the ability to add stream, datasets.
+ */
+public interface DatasetConfigurer {
+
+  /**
+   * Adds a {@link Stream}.
+   *
+   * @param stream {@link Stream}
+   */
+  void addStream(Stream stream);
+
+  /**
+   * Adds a {@link DatasetModule} to be deployed automatically (if absent in the CDAP instance) during application
+   * deployment.
+   *
+   * @param moduleName Name of the module to deploy
+   * @param moduleClass Class of the module
+   */
+  @Beta
+  void addDatasetModule(String moduleName, Class<? extends DatasetModule> moduleClass);
+
+  /**
+   * Adds a {@link DatasetModule} to be deployed automatically (if absent in the CDAP instance) during application
+   * deployment, using {@link Dataset} as a base for the {@link DatasetModule}.
+   * The module will have a single dataset type identical to the name of the class in the datasetClass parameter.
+   *
+   * @param datasetClass Class of the dataset; module name will be the same as the class in the parameter
+   */
+  @Beta
+  void addDatasetType(Class<? extends Dataset> datasetClass);
+
+  /**
+   * Adds a Dataset instance, created automatically if absent in the CDAP instance.
+   * See {@link co.cask.cdap.api.dataset.DatasetDefinition} for details.
+   *
+   * @param datasetName Name of the dataset instance
+   * @param typeName Name of the dataset type
+   * @param properties Dataset instance properties
+   */
+  @Beta
+  void createDataset(String datasetName, String typeName, DatasetProperties properties);
```

Comment:
Shall we also have variant that doesn't take `DatasetProperties` so that caller doesn't need to pass in `DatasetProperties.EMPTY`?